Newsletters: add 201 (2022-05-25) (#770)
bitcoinops/bitcoinops.github.io: _posts/en/newsletters/2022-05-25-newsletter.md

---
title: 'Bitcoin Optech Newsletter #201'
permalink: /en/newsletters/2022/05/25/
name: 2022-05-25-newsletter
slug: 2022-05-25-newsletter
type: newsletter
layout: newsletter
lang: en
---
This week's newsletter summarizes a draft BIP for package relay and
provides an overview of a concern with Miner Extractable Value (MEV) for
Bitcoin covenant design.  Also included are our regular sections with
the summary of top questions and answers from the Bitcoin Stack
Exchange, announcements of new releases and release candidates, and
descriptions of notable changes to popular Bitcoin infrastructure
software.

## News

- **Package relay proposal:** Gloria Zhao [posted][zhao package] to the
  Bitcoin-Dev mailing list a draft BIP for [package relay][topic package
  relay].  This is a feature that can make [CPFP fee bumping][topic
  cpfp] much more reliable at ensuring a child transaction can
  contribute fees towards getting its parent confirmed.  Several
  contract protocols, including LN, already require reliable CPFP fee

>   bumping, so package relay would improve the security and usability of
>   those protocols.  The draft BIP proposes to add four new messages to

Reviewer: ```suggestion
  bumping, so package relay would improve their security and usability.
  The draft BIP proposes to add four new messages to
```

  - `pckginfo1` provides information about a package of transactions
    including the number of transactions, each transaction's identifier
    (wtxid), the total size (weight) of the transactions, and
    the total fees of the transaction.  The feerate of the package can be
    calculated by dividing fees by weight.

  Additionally the existing `inv` and `getdata` messages are updated
  with a new inventory (inv) type, `MSG_PCKG1`, which allows a node to
  tell its peers that it is willing to send a `pckginfo1` message about
  a transaction and which those peers can then use to request that
  `pckginfo1` message for a particular transaction.

  Using these messages, a node can use an `inv(MSG_PCKG1)` message to tell
  its peers that they might be interested in receiving a `pckginfo1` about
  a transaction, e.g. because it is a high-feerate child of a low-feerate
  unconfirmed parent transaction the peers might otherwise ignore.  If any
  peer requests the `pckginfo1` message, they can use the info in the
  message to determine whether they really are interested in the package
  and to learn the wtxids of any transactions they need to download in

>   order to validate the high feerate child.  The actually transactions can

Reviewer: ```suggestion
  order to validate the high feerate child.  The actual transactions can
```

  be requested with a `getpckgtxns` message and received in a `pckgtxins`
  message.

  Although the [draft BIP][bip-package-relay] focuses on just the
  protocol, Zhao's email provides additional context, briefly describes

>   alternative designs that were considered by found deficient, and links

Reviewer: ```suggestion
  alternative designs that were found deficient, and links
```

  to a [presentation][zhao preso] with additional details.

- **Miner extractable value discussion:** developer /dev/fd0
  [posted][fd0 ctv9] to the Bitcoin-Dev mailing list a summary of the
  [ninth IRC meeting][ctv9] about [OP_CHECKTEMPLATEVERIFY][topic
  op_checktemplateverify] (CTV).  Among other topics discussed during
  that meeting, Jeremy Rubin listed several concerns he's heard people
  express about recursive [covenants][topic covenants] (which CTV does
  not enable).  One of those concerns was creating Miner Extractable
  Value (MEV) significantly beyond the amount available from running a
  simple transaction selection algorithm such as that provided by
  Bitcoin Core.

    MEV has become a particular concern in Ethereum and related
    protocols where the use of public onchain trading protocols allows
    miners to frontrun trades.  For example, imagine the following two
    unconfirmed transactions both being available to mine in the next
    block:

>     * Alice sells *x* to Bob for 1 ETH

Reviewer 2: what is x?

    * Bob sells *x* to Carol for 2 ETH (Bob earns 1 ETH in profit)

    <br>If those two exchanges are conducted using a public onchain trading
    protocol, a miner can cut Bob out of the transaction.  For example:

    * Alice sells *x* to Miner Mallory for 1 ETH
    * Miner Mallory sells *x* to Carol for 2 ETH (Mallory earns 1 ETH in
      profit; Bob earns nothing)

    <br>Obviously this is a problem for Bob, but it also creates several
    problems for the network.  The first problem is that, in the above
    example, Miner Mallory needs to have 1 ETH in a wallet in order to
    exploit the situation.  Every other miner attempting to mine the

>     next block also needs to have their own 1 ETH.  That wallet value is

Reviewer 2: this is not really a problem in Eth because flashloans exist

    independent of each miner's hashrate, so two miners can join
    together to halve the amount of money they each need to store in
    their wallets for capturing MEV---producing an economy of scale that
    encourages centralization of mining and which can leave the network

>     more vulnerable to transaction censorship.  A [report][bitmex

Reviewer 2: Again, flashloans solves this in a decentralized way :)

Reviewer 3: Isn't ethereum's flashbots mechanism entirely centralized though?

From the Bitmex report:

> Flashbots is a single centralised entity. [..] Searchers send MEV transactions to the server with the address https://relay.flashbots.net/

Reviewer 2: flash loans is entirely separate from flashbots.

Reviewer 3: Oh right I see what you mean. Flashloans can lower the capital requirements for exploiting the arbitrage opportunity, even lacking flashbots.

But can flashloans be used reliably lacking a flashbots-like protocol that provides a (semi)private communication channel between searchers and miners? Won't other network participants just 'steal' your flashloan strategy and resubmit it with a higher mining fee otherwise?

I fail to see how something like this could be both stable and decentralized, but maybe I'm just not imaginative enough :)

Reviewer 2: The way that flashloans work is that you just post capital into a contract and anyone is allowed to borrow it and return it atomically, with some interest for the loan.

Flashbots can either use a flashloan or not use one, depending on the strategy. I don't think flashbots does have their own flashloans pool...

instead, you want to do something like borrow from uniswap https://docs.uniswap.org/protocol/V2/guides/smart-contract-integration/using-flash-swaps, because the capital sitting there is also "productive" for uniswap market making, and whatever else.

    flashbots] by BitMex Research claims that a centralized service that
    brokers these types of MEV transactions was being used by 90% of

>     Ethereum hashrate at the time the report was written.  To obtain the

Reviewer 2: Flashbots is doing something else actually, it's not about capital. It's about the "searchers" and strategy makers coordinating with the miners to get paid even if they don't mine.

    maximum returns, that service could be changed to discourage mining

>     competing transactions, which would effective give it the power to

Reviewer: ```suggestion
    competing transactions, effectively giving it the power to
```

    censor transactions if it was used by 100% of miners (or if it was
    used by more than 50% of miners and allowed to engage in block chain
    reorganization).

    A second problem occurs if the original transaction from Alice to
    Bob is confirmed in a block before the transaction from Bob to Carol

>     becomes available.  Now there's a 1 ETH bounty for any miner who

Reviewer 2: this is sort of confusing and not really the problem.

The issue is that Mallory's modified txn is also reorgable by Eve.

    re-mines the previous block.  This pressure to re-mine blocks
    exacerbates [fee sniping][topic fee sniping] pressure, which at its
    worst can make confirmation scores useless for determining
    transaction finality, eliminating the ability to use proof of work

>     to secure the network.

Reviewer 2: slightly worse here:

the amount of *energy* invested into potentially NP-hard MEV searches might even compete with bitcoin PoW energy usage.

Author: Why is that a problem?  It's ok if more energy than Bitcoin PoW is expended on moving vehicles, or powering homes and offices, or even on traditional finance.  So it should also be ok if more energy is invested into finding defi arbitrage opportunities than is expended on PoW.  I think the problems occur when:

1. Someone is producing PoW for deliberate reorgs
2. MEV searching is subject to strong economies of scale because, if MEV provides significant profits, mining becomes subject to strong economies of scale, which is centralizing

Author: Right, I agree about lucrative reorging being bad for consensus stability.  I'm still trying to understand why it matters that the chips and energy used for it might compete with Bitcoin PoW mining.  Is it just that the amount of demand for MEV finding might be highly dynamic, so during periods of high demand for MEV finding, prices for energy for PoW might spike and cause fewer blocks to be produced?  Extending that out, inconsistencies in block production might cause dips and spikes in regular transaction fees, which might make fee sniping more profitable, which might further undermine consensus stability?

{% comment %}<!-- https://bitcoin.stackexchange.com/search?tab=votes&q=created%3a1m..%20is%3aanswer -->{% endcomment %}
{% assign bse = "https://bitcoin.stackexchange.com/a/" %}

- [How much entropy is lost alphabetising your mnemonics?]({{bse}}113432)
  HansBKK wonders how much entropy is lost if 12 or 24 word seed phrases were sorted
  alphabetically. Pieter Wuille breaks down a series of metrics including number
  of possibilities, entropy, and mean number of guesses to brute force across 12 and 24
  words and also notes considerations for word repetition.

- [Taproot signing with PSBT: How to determine signing method?]({{bse}}113489)
  Guggero lists three ways of providing a valid [Schnorr signature][topic
  schnorr signatures] in taproot: a keypath spend with [BIP86][] commitment, a
  keypath spend with a commitment to the root of the script tree, and a
  scriptpath spend. Andrew Chow confirms Guggero's outline of how each signing
  method is indicated within a [PSBT][topic psbt].

- [How would faster blocks cause mining centralization?]({{bse}}113505)

>   Murch focuses on why shorter block times would cause more frequent
>   reorganizations and benefit larger miners and how block propagation latency
>   plays a role.

Reviewer 4: The double and was a bit strange. How do you like:

```suggestion
  Murch focuses on why shorter block times would cause more frequent
  reorganizations and how those would benefit larger miners in the context 
  of block propagation latency.
```
